apps/enc.c

```diff
@@ -418,15 +426,31 @@ int enc_main(int argc, char **argv)
                     BIO_printf(bio_err, "bad magic number\n");
                     goto end;
                 }
-
                 sptr = salt;
             }
 
-            if (!EVP_BytesToKey(cipher, dgst, sptr,
-                                (unsigned char *)str,
-                                str_len, 1, key, iv)) {
-                BIO_printf(bio_err, "EVP_BytesToKey failed\n");
-                goto end;
+            if (pbkdf2 == 1) {
+                /* generate key and default iv concatenated in a temporary buffer */
+                unsigned char tmpkeyiv[EVP_MAX_KEY_LENGTH + EVP_MAX_IV_LENGTH];
+                int iKlen = EVP_CIPHER_key_length(cipher);
+                int iVlen = EVP_CIPHER_iv_length(cipher);
+                int iSlen = sptr ? sizeof(salt) : 0;  /* not needed if HASH_UPDATE() is fixed */
+                if (!PKCS5_PBKDF2_HMAC(str, str_len, sptr, iSlen, iter,
+                                       dgst,
+                                       iKlen+iVlen, tmpkeyiv)) {
+                    BIO_printf(bio_err, "PKCS5_PBKDF2_HMAC failed\n");
+                    goto end;
+                }
+                /* split and move data back to global buffer */
+                memcpy(key, tmpkeyiv, iKlen);
+                memcpy(iv, tmpkeyiv+iKlen, iVlen);
+            } else {
+                if (!EVP_BytesToKey(cipher, dgst, sptr,
+                                    (unsigned char *)str,
+                                    str_len, 1, key, iv)) {
+                    BIO_printf(bio_err, "EVP_BytesToKey failed\n");
+                    goto end;
```

Comment:
I think I lost the argument about making PKCS5_PBKDF2_HMAC the default following internal discussion because it breaks compatibility. However I do think we should print a warning message along with a hint to use the "-iter" option (or -pbkdf2 if we decide to add that) if we end up using this algorithm